### ISSUE TITLE
UV for obj exporting is broken for the latest meshlab 2016

### ISSUE THREAD

**Issue by jasjuang** [cnr-isti-vclab/meshlab#67]
I am on Windows 10, using the latest meshlab 2016. If I import a random mesh with UV information and export it as an obj, the UV coordinates will be assigned 0 for all of them. I can confirm that all the TexCoord checkbox is checked. Also this problem does not exist in the old meshlab. Can someone please fix it?

**Comment by mcallieri** [comment]
I just checked, the exporter works correctly. 
However, remember that UV coordinates may be PER VERTEX or PER WEDGE. You an see which kind UV you have in the info panel (the red strip in te 3D window): VT means vertex UV, while WT means wedge UV. You may have both at the same time. 
It is possible to convert from VT to WT using the filters "convert PerVertex UV into PerWedge UV" and "convert PerWedge UV into PerVertex UV", under filters->texture.
When you export, you may export VT and/or WT (there are two checkboxes "TexCoords", one for vertex and one for wedge), remember to check/uncheck the one you want.

**Comment by jasjuang** [comment]
I created a minimal cube obj to recreate this problem for you to test as well

``````
v -0.500000 -0.500000 0.500000
v 0.500000 -0.500000 0.500000
v -0.500000 0.500000 0.500000
v 0.500000 0.500000 0.500000
v -0.500000 0.500000 -0.500000
v 0.500000 0.500000 -0.500000
v -0.500000 -0.500000 -0.500000
v 0.500000 -0.500000 -0.500000
vt 0.375000 0.000000
vt 0.625000 0.000000
vt 0.375000 0.250000
vt 0.625000 0.250000
vt 0.375000 0.500000
vt 0.625000 0.500000
vt 0.375000 0.750000
vt 0.625000 0.750000
vt 0.375000 1.000000
vt 0.625000 1.000000
vt 0.875000 0.000000
vt 0.875000 0.250000
vt 0.125000 0.000000
vt 0.125000 0.250000
vn 0.000000 0.000000 1.000000
vn 0.000000 0.000000 1.000000
vn 0.000000 0.000000 1.000000
vn 0.000000 0.000000 1.000000
vn 0.000000 1.000000 0.000000
vn 0.000000 1.000000 0.000000
vn 0.000000 1.000000 0.000000
vn 0.000000 1.000000 0.000000
vn 0.000000 0.000000 -1.000000
vn 0.000000 0.000000 -1.000000
vn 0.000000 0.000000 -1.000000
vn 0.000000 0.000000 -1.000000
vn 0.000000 -1.000000 0.000000
vn 0.000000 -1.000000 0.000000
vn 0.000000 -1.000000 0.000000
vn 0.000000 -1.000000 0.000000
vn 1.000000 0.000000 0.000000
vn 1.000000 0.000000 0.000000
vn 1.000000 0.000000 0.000000
vn 1.000000 0.000000 0.000000
vn -1.000000 0.000000 0.000000
vn -1.000000 0.000000 0.000000
vn -1.000000 0.000000 0.000000
vn -1.000000 0.000000 0.000000
s off
f 1/1/1 2/2/2 4/4/3 3/3/4
f 3/3/5 4/4/6 6/6/7 5/5/8
f 5/5/9 6/6/10 8/8/11 7/7/12
f 7/7/13 8/8/14 2/10/15 1/9/16
f 2/2/17 8/11/18 6/12/19 4/4/20
f 7/13/21 1/1/22 3/3/23 5/14/24
``````

When I re-export this with meshlab 2016 with the TexCoord checkbox checked here is the result, the TexCoord checkbox on the vertex section is blacked out so I can't check it and the TexCoord checkbox on the Wedge section is checked. It's clearly vt at the original file but I have no idea why meshlab is classifying it as wt

``````
####
#
# OBJ File Generated by Meshlab
#
####
# Object After.obj
#
# Vertices: 8
# Faces: 6
#
####
mtllib ./After.obj.mtl

v -0.500000 -0.500000 0.500000
v 0.500000 -0.500000 0.500000
v -0.500000 0.500000 0.500000
v 0.500000 0.500000 0.500000
v -0.500000 0.500000 -0.500000
v 0.500000 0.500000 -0.500000
v -0.500000 -0.500000 -0.500000
v 0.500000 -0.500000 -0.500000
# 8 vertices, 0 vertices normals


usemtl material_0
vt 0.000000 0.000000
vt 0.000000 0.000000
f 2/1 4/1 3/1 1/2
vt 0.000000 0.000000
f 4/1 6/1 5/1 3/3
f 6/1 8/1 7/1 5/3
f 8/1 2/1 1/1 7/3
f 8/1 6/1 4/1 2/3
f 1/1 3/1 5/1 7/1
# 6 faces, 3 coords texture

# End of File
``````

Notice that the vt are set to 0, which is an error. For your reference, if I do the exact same re-exporting in the old meshlab, here is the result

``````
####
#
# OBJ File Generated by Meshlab
#
####
# Object test.obj
#
# Vertices: 8
# Faces: 12
#
####
mtllib ./test.obj.mtl

v -0.500000 -0.500000 0.500000
v 0.500000 -0.500000 0.500000
v -0.500000 0.500000 0.500000
v 0.500000 0.500000 0.500000
v -0.500000 0.500000 -0.500000
v 0.500000 0.500000 -0.500000
v -0.500000 -0.500000 -0.500000
v 0.500000 -0.500000 -0.500000
# 8 vertices, 0 vertices normals


usemtl material_0
vt 0.375000 0.000000
vt 0.625000 0.000000
vt 0.625000 0.250000
f 1/1 2/2 4/3
vt 0.375000 0.250000
f 1/1 4/3 3/4
vt 0.625000 0.500000
f 3/4 4/3 6/5
vt 0.375000 0.500000
f 3/4 6/5 5/6
vt 0.625000 0.750000
f 5/6 6/5 8/7
vt 0.375000 0.750000
f 5/6 8/7 7/8
vt 0.625000 1.000000
f 7/8 8/7 2/9
vt 0.375000 1.000000
f 7/8 2/9 1/10
vt 0.875000 0.000000
vt 0.875000 0.250000
f 2/2 8/11 6/12
f 2/2 6/12 4/3
vt 0.125000 0.000000
f 7/13 1/1 3/4
vt 0.125000 0.250000
f 7/13 3/4 5/14
# 12 faces, 14 coords texture

# End of File
``````

It's ordered a little bit differently then the original file but it is working correctly.

**Comment by mcallieri** [comment]
First of all, in this example texture coordinated ARE per-wedge and not per-vertex, as each quad has 4 different UV coords, and each vertex, depending on which face is on, has different UV.

Then, yes, the exporter is messing up UV coordinates. It seems to happen only when non-triangular faces are involved. MeshLab always convert polygons into triangles, as it does not work natively with n-gons.
We will correct the problem, and let you know when it has been solved.

**Comment by jasjuang** [comment]
You are right. I did some testing myself and the error only happens when the face is not a triangle. Look forward to the fix!

**Comment by mcallieri** [comment]
We have isolated the problem. MeshLab uses triangles, but it also can store non-triangular polygons using faux-edges. If we know the mesh has faux-edges (it was imported from a file with non-triangle polygon, or was generated by some filters) we can try to convert from triangles to faces, before exporting. This procedure, however, is not completely correct, and has problems with some data like per-wedge UV.
So, as a temporary fix, we made this triangle-to-polygon convertion optional (there is now a parameter when exporting). So, in the example above, loading and then saving no longer destroys UV coordinates, BUT the faces are converted to triangles.
We hope to have a proper fix (a correct conversion from triangles to polygons) sooner or later, but in the meantime we can do this.
It will be available in the next release (around 1 month).
